test/EFCore.PG.FunctionalTests/Query/HstoreQueryTest.cs

```diff
@@ -0,0 +1,255 @@
+    {
+        var keyToTest = "key2";
+        await AssertQuery(async, _ => _.Set<DictionaryEntity>().Where(s => s.NullableDictionary != null && s.NullableDictionary.ContainsKey(keyToTest)));
+        AssertSql("""
+@__keyToTest_0='key2'
+
+SELECT s."Id", s."Dictionary", s."DictionaryContainerEntityId", s."ImmutableDictionary", s."NullableDictionary", s."NullableImmutableDictionary"
+FROM "SomeEntities" AS s
+WHERE s."NullableDictionary" IS NOT NULL AND s."NullableDictionary" ? @__keyToTest_0
+""");
+    }
+
+    [Theory]
+    [MemberData(nameof(IsAsyncData))]
+    public async Task ImmutableNullableDictionary_Contains_key(bool async)
+    {
+        var keyToTest = "key3";
+        await AssertQuery(async, _ => _.Set<DictionaryEntity>().Where(s => s.NullableImmutableDictionary != null && s.NullableImmutableDictionary.ContainsKey(keyToTest)));
+        AssertSql("""
+@__keyToTest_0='key3'
+
+SELECT s."Id", s."Dictionary", s."DictionaryContainerEntityId", s."ImmutableDictionary", s."NullableDictionary", s."NullableImmutableDictionary"
+FROM "SomeEntities" AS s
+WHERE s."NullableImmutableDictionary" IS NOT NULL AND s."NullableImmutableDictionary" ? @__keyToTest_0
+""");
+    }
+
+    [Theory]
+    [MemberData(nameof(IsAsyncData))]
+    public async Task ImmutableDictionary_Contains_key(bool async)
+    {
+        var keyToTest = "key3";
+        await AssertQuery(async, _ => _.Set<DictionaryEntity>().Where(s => s.ImmutableDictionary.ContainsKey(keyToTest)));
+        AssertSql(
+            """
+@__keyToTest_0='key3'
+
+SELECT s."Id", s."Dictionary", s."DictionaryContainerEntityId", s."ImmutableDictionary", s."NullableDictionary", s."NullableImmutableDictionary"
+FROM "SomeEntities" AS s
+WHERE s."ImmutableDictionary" ? @__keyToTest_0
+""");
+    }
+
+    [Theory]
+    [MemberData(nameof(IsAsyncData))]
+    public async Task Dictionary_Contains_value(bool async)
+    {
+        var valueToTest = "value";
+        await AssertQuery(async, _ => _.Set<DictionaryEntity>().Where(s => s.Dictionary.ContainsValue(valueToTest)));
+        AssertSql(
+            """
+@__valueToTest_0='value'
+
+SELECT s."Id", s."Dictionary", s."DictionaryContainerEntityId", s."ImmutableDictionary", s."NullableDictionary", s."NullableImmutableDictionary"
+FROM "SomeEntities" AS s
+WHERE @__valueToTest_0 = ANY(avals(s."Dictionary"))
+""");
+    }
+
+    [Theory]
+    [MemberData(nameof(IsAsyncData))]
+    public async Task Dictionary_Item_equals_found(bool async)
+    {
+        var keyToTest = "key";
+        var valueToTest = "value";
+        await AssertQuery(async, _ => _.Set<DictionaryEntity>().Where(s => s.Dictionary[keyToTest] == valueToTest));
+        AssertSql(
+            """
+@__valueToTest_0='value'
+
+SELECT s."Id", s."Dictionary", s."DictionaryContainerEntityId", s."ImmutableDictionary", s."NullableDictionary", s."NullableImmutableDictionary"
+FROM "SomeEntities" AS s
+WHERE s."Dictionary" -> 'key' = @__valueToTest_0
```

Comment:
Comparing this line to line 186, you see that when using a normal `Dictionary<string, string>`, accessing the indexer property causes the key being passed as the index to be a constant SQL expression while if use use an `ImmutableDictionary<string, string>, the key being passed as the index is instead a query parameter.

With the `Nullable` dictionary examples, you see the same behavior however you can see that for the `ContainsKey` translation, the key is being passed as the query parameter both in the normal Dictionary and the ImmutableDictionary.

I have wracked my brain and looked throughout the codebase on how this could possibly happen. I do not treat the two types of dictionaries any differently in the hstore code I've added so the variation seems extremely perplexing to me.

This technically will not cause a functional issue.  It will work either way.

However, because the `Dictionary<string, string>` access is being passed as a constant and not a parameter, if a 2nd query is made on a different key in the same running application, it will require a completely new query plan and not reuse the existing cached query.

Anyone have any ideas why this might be happening?

Reply:
It's very understandable that you got stuck on this :) I took a look, and indeed, very early in the EF query pipeilne, at the place where LINQ parameters (captured closure variables) get parameterized, there's an exception for arguments of indexers. I'm not sure why that is - I've opened https://github.com/dotnet/efcore/issues/34733 to look into it at some point.

As you say, this isn't critical, though it indeed has quite a performance cost - every lookup will cause the query to have a cache miss, making it be fully compiled again and again. This isn't great.